File: test/python_http_server_integration.py
#!/usr/bin/env python3
"""Python HTTP server over RSP integration test using the Python client."""

import asyncio
import json
import os
import signal
import sys
from pathlib import Path

HARNESS_TIMEOUT_S = 60.0
ENDORSEMENT_SUCCESS = 0
ETYPE_ACCESS = "f1e2d3c4-5b6a-7980-1a2b-3c4d5e6f7a8b"
EVALUE_ACCESS_NETWORK = "f1e2d3c4-5b6a-7980-1a2b-3c4d5e6f7a8b"
ETYPE_ROLE = "0963c0ab-215f-42c1-b042-747bf21e330e"
EVALUE_ROLE_CLIENT = "edab2025-4ae1-44f2-a683-1a390586e10c"
EVALUE_ROLE_RESOURCE_SERVICE = "a7f8c9d6-3b2e-4f1a-8c9d-5e6f7a8b9c0d"


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[1]


CLIENT_PYTHON_DIR = _repo_root() / "client" / "python"
if str(CLIENT_PYTHON_DIR) not in sys.path:
    sys.path.insert(0, str(CLIENT_PYTHON_DIR))

from rsp_client import RSPClient  # noqa: E402
from rsp_net import create_connection  # noqa: E402


async def with_timeout(coro, timeout_s: float, label: str):
    try:
        return await asyncio.wait_for(coro, timeout=timeout_s)
    except asyncio.TimeoutError as exc:
        raise RuntimeError(f"{label} timed out after {timeout_s}s") from exc


async def _drain_stream(stream: asyncio.StreamReader, sink: list[str], prefix: str) -> None:
    while True:
        line = await stream.readline()
        if not line:
            return
        text = line.decode("utf-8", errors="replace").rstrip()
        sink.append(text)
        print(f"[{prefix}] {text}", file=sys.stderr)


async def wait_for_fixture_ready(proc: asyncio.subprocess.Process, stderr_lines: list[str]) -> dict:
    stdout_lines = []
    while proc.returncode is None:
        line = await proc.stdout.readline()
        if not line:
            break
        text = line.decode("utf-8", errors="replace").strip()
        stdout_lines.append(text)
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            continue

    await proc.wait()
    raise RuntimeError(
        "fixture exited before readiness JSON"
        + (f"\nstdout:\n{os.linesep.join(stdout_lines)}" if stdout_lines else "")
        + (f"\nstderr:\n{os.linesep.join(stderr_lines)}" if stderr_lines else "")
    )


async def wait_for_server_ready(proc: asyncio.subprocess.Process) -> None:
    while True:
        line = await proc.stdout.readline()
        if not line:
            stderr_output = b""
            if proc.stderr is not None:
                try:
                    stderr_output = await asyncio.wait_for(proc.stderr.read(), timeout=0.5)
                except Exception:
                    stderr_output = b""
            raise RuntimeError(
                "python server exited before ready"
                + (f"\nstderr:\n{stderr_output.decode('utf-8', errors='replace')}" if stderr_output else "")
            )
        text = line.decode("utf-8", errors="replace").strip()
        if "Python HTTP server over RSP listening on" in text:
            return


async def read_until_expected(socket, expected_tokens: list[str], timeout_s: float) -> str:
    deadline = asyncio.get_running_loop().time() + timeout_s
    chunks = bytearray()
    while True:
        text = chunks.decode("utf-8", errors="replace")
        if all(token in text for token in expected_tokens):
            return text

        remaining = deadline - asyncio.get_running_loop().time()
        if remaining <= 0:
            raise RuntimeError("timed out waiting for expected HTTP response tokens")

        chunk = await asyncio.wait_for(socket.read(), timeout=min(1.0, remaining))
        if not chunk:
            return chunks.decode("utf-8", errors="replace")
        chunks.extend(chunk)


async def request_endorsement_or_throw(client: RSPClient, endorsement_node_id: str, endorsement_type: str, endorsement_value: str, label: str) -> None:
    value_bytes = endorsement_value.encode("utf-8")
    reply = None
    for _ in range(3):
        reply = await client.begin_endorsement_request(endorsement_node_id, endorsement_type, value_bytes)
        if reply and reply.get("status") == ENDORSEMENT_SUCCESS:
            return
    status = reply.get("status") if isinstance(reply, dict) else "null"
    raise RuntimeError(f"{label} endorsement failed (status={status})")


async def terminate_process(proc: asyncio.subprocess.Process) -> None:
    if proc is None or proc.returncode is not None:
        return
    proc.terminate()
    try:
        await asyncio.wait_for(proc.wait(), timeout=2.0)
    except asyncio.TimeoutError:
        proc.kill()
        await proc.wait()


async def main() -> None:
    if len(sys.argv) < 2:
        raise RuntimeError("Usage: python3 test/python_http_server_integration.py <fixture-path>")

    fixture_path = sys.argv[1]
    fixture = await asyncio.create_subprocess_exec(
        fixture_path,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    fixture_stderr_lines: list[str] = []
    fixture_stderr_task = asyncio.create_task(_drain_stream(fixture.stderr, fixture_stderr_lines, "fixture:stderr"))

    server_proc = None
    server_stderr_task = None
    server_stderr_lines: list[str] = []
    requester_client = None

    try:
        print("[python_http_server_integration] waiting for fixture", file=sys.stderr)
        info = await with_timeout(wait_for_fixture_ready(fixture, fixture_stderr_lines), 30.0, "fixture readiness")
        print("[python_http_server_integration] fixture ready", file=sys.stderr)

        host_port = "127.0.0.1:51880"
        app_path = _repo_root() / "integration" / "python_http_server" / "working" / "app.py"

        env = os.environ.copy()
        env.update({
            "RSP_TRANSPORT": info["transport_spec"],
            "RSP_RESOURCE_SERVICE_NODE_ID": info["resource_service_node_id"],
            "RSP_ENDORSEMENT_NODE_ID": info["endorsement_service_node_id"],
            "RSP_HOST_PORT": host_port,
        })

        server_proc = await asyncio.create_subprocess_exec(
            sys.executable,
            str(app_path),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=env,
        )

        server_stderr_task = asyncio.create_task(_drain_stream(server_proc.stderr, server_stderr_lines, "python-server:stderr"))

        print("[python_http_server_integration] waiting for python server", file=sys.stderr)
        await with_timeout(wait_for_server_ready(server_proc), 25.0, "python server readiness")
        print("[python_http_server_integration] python server ready", file=sys.stderr)

        requester_client = RSPClient()
        print("[python_http_server_integration] connecting requester", file=sys.stderr)
        await with_timeout(requester_client.connect(info["transport_spec"]), 5.0, "requester connect")
        print("[python_http_server_integration] requester connected", file=sys.stderr)

        reachable = await with_timeout(
            requester_client.ping(info["endorsement_service_node_id"], timeout=3.0),
            4.0,
            "endorsement service ping",
        )
        if not reachable:
            raise RuntimeError("endorsement service unreachable from requester client")
        print("[python_http_server_integration] endorsement ping ok", file=sys.stderr)

        await request_endorsement_or_throw(
            requester_client,
            info["endorsement_service_node_id"],
            ETYPE_ACCESS,
            EVALUE_ACCESS_NETWORK,
            "network access",
        )
        await request_endorsement_or_throw(
            requester_client,
            info["endorsement_service_node_id"],
            ETYPE_ROLE,
            EVALUE_ROLE_CLIENT,
            "client role",
        )
        await request_endorsement_or_throw(
            requester_client,
            info["endorsement_service_node_id"],
            ETYPE_ROLE,
            EVALUE_ROLE_RESOURCE_SERVICE,
            "resource service role",
        )

        socket = await with_timeout(
            create_connection(requester_client, info["resource_service_node_id"], host_port),
            6.0,
            "create rsp connection",
        )
        print("[python_http_server_integration] rsp socket connected", file=sys.stderr)

        request_bytes = (
            b"GET / HTTP/1.1\r\n"
            b"Host: python-rsp.local\r\n"
            b"Connection: close\r\n\r\n"
        )
        await with_timeout(socket.write(request_bytes), 4.0, "http request write")

        response = await with_timeout(
            read_until_expected(socket, ["HTTP/1.1 200 OK", "Remote Socket Protocol"], 6.0),
            7.0,
            "http response read",
        )
        await socket.close()
        print("[python_http_server_integration] response received", file=sys.stderr)

        assert "HTTP/1.1 200 OK" in response
        print("python_http_server_integration passed")
    finally:
        print("[python_http_server_integration] cleanup begin", file=sys.stderr)
        if requester_client is not None:
            await requester_client.close()
        if server_proc is not None:
            await terminate_process(server_proc)
        if server_stderr_task is not None:
            server_stderr_task.cancel()
            await asyncio.gather(server_stderr_task, return_exceptions=True)
        await terminate_process(fixture)
        fixture_stderr_task.cancel()
        await asyncio.gather(fixture_stderr_task, return_exceptions=True)
        print("[python_http_server_integration] cleanup done", file=sys.stderr)


if __name__ == "__main__":
    async def _entrypoint() -> None:
        await asyncio.wait_for(main(), timeout=HARNESS_TIMEOUT_S)

    try:
        asyncio.run(_entrypoint())
    except asyncio.TimeoutError:
        print(f"python_http_server_integration failed: harness timed out after {HARNESS_TIMEOUT_S}s", file=sys.stderr)
        raise SystemExit(1)

File: test/python_http_server_integration.js
'use strict';

const assert = require('assert');
const path = require('path');
const readline = require('readline');
const {spawn} = require('child_process');

const {RSPClient} = require('../client/nodejs/rsp_client');
const {createConnection} = require('../client/nodejs/rsp_net');

const ENDORSEMENT_SUCCESS = 0;
const ETYPE_ACCESS = 'f1e2d3c4-5b6a-7980-1a2b-3c4d5e6f7a8b';
const EVALUE_ACCESS_NETWORK = 'f1e2d3c4-5b6a-7980-1a2b-3c4d5e6f7a8b';
const ETYPE_ROLE = '0963c0ab-215f-42c1-b042-747bf21e330e';
const EVALUE_ROLE_CLIENT = 'edab2025-4ae1-44f2-a683-1a390586e10c';
const EVALUE_ROLE_RESOURCE_SERVICE = 'a7f8c9d6-3b2e-4f1a-8c9d-5e6f7a8b9c0d';
const HARNESS_TIMEOUT_MS = 20000;

function withTimeout(promise, timeoutMs, label) {
    return Promise.race([
        promise,
        new Promise((_, reject) => {
            setTimeout(() => reject(new Error(`${label} timed out after ${timeoutMs}ms`)), timeoutMs);
        }),
    ]);
}

function waitForFixtureReady(fixture) {
    return new Promise((resolve, reject) => {
        const stdoutLines = [];
        const stderrLines = [];
        const stdoutReader = readline.createInterface({input: fixture.stdout});
        const stderrReader = readline.createInterface({input: fixture.stderr});

        stdoutReader.on('line', (line) => {
            stdoutLines.push(line);
            try {
                resolve({info: JSON.parse(line), stdoutLines, stderrLines});
            } catch {
            }
        });

        stderrReader.on('line', (line) => {
            stderrLines.push(line);
            console.error(`[python_http_server_integration][fixture] ${line}`);
        });

        fixture.once('error', (error) => reject(error));
        fixture.once('exit', (code, signal) => {
            reject(new Error(
                `fixture exited before ready (code=${code}, signal=${signal})\nstdout:\n${stdoutLines.join('\n')}\nstderr:\n${stderrLines.join('\n')}`
            ));
        });
    });
}

function waitForPythonServerReady(proc) {
    return new Promise((resolve, reject) => {
        const stdoutReader = readline.createInterface({input: proc.stdout});
        const stderrLines = [];
        const stderrReader = readline.createInterface({input: proc.stderr});

        stdoutReader.on('line', (line) => {
            if (line.includes('Python HTTP server over RSP listening on')) {
                resolve();
            }
        });

        stderrReader.on('line', (line) => {
            stderrLines.push(line);
            console.error(`[python_http_server_integration][python] ${line}`);
        });

        proc.once('error', (error) => reject(error));
        proc.once('exit', (code, signal) => {
            reject(new Error(
                `python server exited before ready (code=${code}, signal=${signal})\nstderr:\n${stderrLines.join('\n')}`
            ));
        });
    });
}

async function requestEndorsementOrThrow(client, endorsementNodeId, endorsementType, endorsementValue, label) {
    let reply = null;
    for (let attempt = 0; attempt < 3 && (!reply || reply.status !== ENDORSEMENT_SUCCESS); attempt += 1) {
        reply = await withTimeout(
            client.beginEndorsementRequest(endorsementNodeId, endorsementType, endorsementValue),
            3000,
            `${label} endorsement request`
        );
    }
    if (!reply || reply.status !== ENDORSEMENT_SUCCESS) {
        throw new Error(`${label} endorsement failed (status=${reply ? reply.status : 'null'})`);
    }
}

function readUntilExpectedWithTimeout(stream, timeoutMs, expectedTokens) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        let resolved = false;

        const timer = setTimeout(() => {
            cleanup();
            reject(new Error(`timed out waiting for expected HTTP response content after ${timeoutMs}ms`));
        }, timeoutMs);

        const maybeResolve = () => {
            if (resolved || chunks.length === 0) {
                return;
            }
            const text = Buffer.concat(chunks).toString('utf8');
            if (expectedTokens.every((token) => text.includes(token))) {
                resolved = true;
                cleanup();
                resolve(text);
            }
        };

        const onData = (chunk) => {
            chunks.push(chunk);
            maybeResolve();
        };

        const onEnd = () => {
            if (!resolved) {
                resolved = true;
                cleanup();
                resolve(Buffer.concat(chunks).toString('utf8'));
            }
        };

        const onError = (error) => {
            if (!resolved) {
                resolved = true;
                cleanup();
                reject(error);
            }
        };

        const cleanup = () => {
            clearTimeout(timer);
            stream.off('data', onData);
            stream.off('end', onEnd);
            stream.off('error', onError);
        };

        stream.on('data', onData);
        stream.on('end', onEnd);
        stream.on('error', onError);
    });
}

async function terminateProcess(proc) {
    if (!proc || proc.exitCode !== null || proc.killed) {
        return;
    }
    proc.kill('SIGTERM');
    await new Promise((resolve) => {
        const timer = setTimeout(() => {
            proc.kill('SIGKILL');
            resolve();
        }, 1500);
        proc.once('exit', () => {
            clearTimeout(timer);
            resolve();
        });
    });
}

async function main() {
    const [, , fixturePath] = process.argv;
    if (!fixturePath) {
        throw new Error('Usage: node test/python_http_server_integration.js <fixture-path>');
    }

    const fixture = spawn(fixturePath, [], {stdio: ['ignore', 'pipe', 'pipe']});
    let pythonServer = null;
    let requesterClient = null;

    try {
        const {info} = await withTimeout(waitForFixtureReady(fixture), 10000, 'fixture readiness');

        const appPath = path.resolve(__dirname, '../integration/python_http_server/working/app.py');
        const pythonBin = path.resolve(__dirname, '../.venv/bin/python');
        const hostPort = '127.0.0.1:51880';

        pythonServer = spawn(pythonBin, [appPath], {
            stdio: ['ignore', 'pipe', 'pipe'],
            env: {
                ...process.env,
                RSP_TRANSPORT: info.transport_spec,
                RSP_RESOURCE_SERVICE_NODE_ID: info.resource_service_node_id,
                RSP_ENDORSEMENT_NODE_ID: info.endorsement_service_node_id,
                RSP_HOST_PORT: hostPort,
            },
        });

        await withTimeout(waitForPythonServerReady(pythonServer), 15000, 'python server readiness');

        requesterClient = new RSPClient();
        await withTimeout(requesterClient.connect(info.transport_spec), 5000, 'requester connect');

        const reachable = await withTimeout(requesterClient.ping(info.endorsement_service_node_id, 3000), 4000, 'endorsement ping');
        if (!reachable) {
            throw new Error('endorsement service unreachable from requester client');
        }

        await requestEndorsementOrThrow(
            requesterClient,
            info.endorsement_service_node_id,
            ETYPE_ACCESS,
            EVALUE_ACCESS_NETWORK,
            'network access'
        );
        await requestEndorsementOrThrow(
            requesterClient,
            info.endorsement_service_node_id,
            ETYPE_ROLE,
            EVALUE_ROLE_CLIENT,
            'client role'
        );
        await requestEndorsementOrThrow(
            requesterClient,
            info.endorsement_service_node_id,
            ETYPE_ROLE,
            EVALUE_ROLE_RESOURCE_SERVICE,
            'resource service role'
        );

        const socket = await withTimeout(
            createConnection(requesterClient, info.resource_service_node_id, hostPort, {asyncData: true}),
            6000,
            'rsp socket connection'
        );

        const responsePromise = readUntilExpectedWithTimeout(socket, 6000, [
            'HTTP/1.1 200 OK',
            'Remote Socket Protocol',
        ]);

        socket.write('GET / HTTP/1.1\r\nHost: python-rsp.local\r\nConnection: close\r\n\r\n');
        const response = await responsePromise;
        assert.ok(response.includes('HTTP/1.1 200 OK'));

        console.log('python_http_server_integration passed');
    } finally {
        const closeRequester = requesterClient ? requesterClient.close().catch(() => {}) : Promise.resolve();
        await Promise.all([
            closeRequester,
            terminateProcess(pythonServer),
            terminateProcess(fixture),
        ]);
    }
}

const watchdog = setTimeout(() => {
    console.error(`python_http_server_integration failed: harness timed out after ${HARNESS_TIMEOUT_MS}ms`);
    process.exit(1);
}, HARNESS_TIMEOUT_MS);

main()
    .then(() => clearTimeout(watchdog))
    .catch((error) => {
        clearTimeout(watchdog);
        console.error(`python_http_server_integration failed: ${error.message}`);
        process.exit(1);
    });
